packages/form-core/src/util-types.ts

```diff
@@ -83,17 +85,40 @@ type PrefixFromDepth<
   TDepth extends any[],
 > = TDepth['length'] extends 0 ? T : `.${T}`
 
-export type DeepValue<TValue, TAccessor> = TValue extends Record<
-  string | number,
-  any
->
-  ? TAccessor extends `${infer TBefore}[${infer TBrackets}].${infer TAfter}`
-    ? DeepValue<TValue[TBefore][TBrackets], TAfter>
-    : TAccessor extends `[${infer TBrackets}]`
-      ? DeepValue<TValue, TBrackets>
-      : TAccessor extends `${infer TBefore}[${infer TBrackets}]`
-        ? DeepValue<TValue[TBefore], TBrackets>
-        : TAccessor extends `${infer TBefore}.${infer TAfter}`
-          ? DeepValue<TValue[TBefore], TAfter>
-          : TValue[TAccessor & string]
-  : never
+export type DeepValue<
+  TValue,
+  TAccessor,
+  TDepth extends any[] = [],
+> = TDepth['length'] extends 3
+  ? never
+  : TValue extends ReadonlyArray<any>
+    ? TAccessor extends `[${infer TBrackets}].${infer TAfter}`
+      ? DeepValue<
+          DeepValue<TValue, TBrackets, [...TDepth, any]>,
+          TAfter,
+          [...TDepth, any]
+        >
+      : TAccessor extends `[${infer TBrackets}]`
+        ? DeepValue<TValue, TBrackets, [...TDepth, any]>
+        : TAccessor extends keyof TValue
+          ? TValue[TAccessor]
+          : TValue[TAccessor & number]
```

Comment:
Could you explain to me why we need `TAccessor &` here:
```suggestion
          : TValue[number]
```

I tried to come up with an answer by myself, but I couldn't. The inferred type here should be the type of the elements in the array, no? 🤔


I ran the tests, with my change, and some of them failed ([source](https://cloud.nx.app/runs/FYlyIB1Kap)):
<img width="1306" alt="image" src="https://github.com/TanStack/form/assets/43729152/31b2194f-747e-4e3d-9201-3fbed38594b9">

I'm a bit lost 😅

Reply:
Funny enough it's more the `& number` that we don't (theoretically) need. TAccessor here is referring to the exact number being passed, while `& number` is type casting it (sorta) to a number so that TS doesn't complain about non numerical access of an array